### DISCUSSION TITLE
How did `userEvent` work before v14 and the `#setup` method?

### DISCUSSION THREAD

**Discussion by stephenpassero** [testing-library/user-event#962] [Q&A] [answered]
I'm installing RTL on a new project, and saw that as of v14, there's this new `#setup` method that is recommended to be used. From the documentation [here](https://testing-library.com/docs/user-event/setup), it looks this "_appl[ies] workarounds and mock[s] the UI layer to simulate user interactions like they would happen in the browser._"

This is all well and good, but how did `userEvent` v13 do this? Did it somehow just not mock the UI layer?

Also kinda a sidenote question, why is everything `async` now?

**Comment by ph-fritsche** [comment, marked as answer]
Previously there wasn't one place for such abstraction but multiple places where workarounds were applied trying to match the browser behavior.
The need for a centralized abstraction of a UI layer is explained https://github.com/testing-library/user-event/issues/714 and further detailed in https://github.com/testing-library/user-event/issues/744 .

As this is open source, you can always [dig through the code (of the previous version)](https://github.com/testing-library/user-event/tree/v13.5.0).

Some APIs were already async before and some new implementations required asynchronous internals.
Always returning a Promise simplifies how we handle those and allows us to change how we handle things internally without further breaking changes. It also eliminates problems with frameworks applying changes asynchronously. For more information see https://github.com/testing-library/user-event/issues/504 and the links in that thread.

**Reply by stephenpassero** [reply]
Ok, thanks! That really helped answer my questions
